types/d3-array/index.d.ts

```diff
@@ -445,312 +491,128 @@ export function descending(a: Primitive | undefined, b: Primitive | undefined):
 // --------------------------------------------------------------------------------------
 
 /**
- * Groups the specified array of values into an InternMap from key to array of value.
- *
- * @param iterable The array to group.
- * @param key The key function.
- */
-export function group<TObject, TKey>(iterable: Iterable<TObject>, key: (value: TObject) => TKey): InternMap<TKey, TObject[]>;
-/**
- * Groups the specified array of values into an InternMap from key to array of value.
- *
- * @param iterable The array to group.
- * @param key1 The first key function.
- * @param key2 The second key function.
- */
-export function group<TObject, TKey1, TKey2>(
-    iterable: Iterable<TObject>,
-    key1: (value: TObject) => TKey1,
-    key2: (value: TObject) => TKey2
-): InternMap<TKey1, InternMap<TKey2, TObject[]>>;
-/**
- * Groups the specified array of values into an InternMap from key to array of value.
+ * Groups the specified iterable of values into an InternMap from key to array of value.
  *
- * @param iterable The array to group.
- * @param key1 The first key function.
- * @param key2 The second key function.
- * @param key3 The third key function.
+ * @param iterable The iterable to group.
+ * @param keys The key functions.
  */
-export function group<TObject, TKey1, TKey2, TKey3>(
+export function group<TObject, TKeys extends Array<(value: TObject, index: number, values: TObject[]) => unknown>>(
```

Comment:
I'm curious if it's possible to flip this on it's head and make it be `TKeys extends unknown[]` and then map the `TKeys` to functions in the parameter.

Reply:
> I'm curious if it's possible to flip this on it's head and make it be `TKeys extends unknown[]` and then map the `TKeys` to functions in the parameter. Seems like that would be a lot easier for people who need to manually specify the type parameters.

@Methuselah96 I will probably find some time on the weekend to take a look at your idea! I will report back to you then

Reply:
I took a quick stab at this and something like this seems promising:

```ts
export type NestedInternMap<
  TObject,
  TReduce,
  TKeys extends unknown[],
> = TKeys extends [infer TFirst, ...infer TRest]
  ? InternMap<TFirst, NestedInternMap<TObject, TReduce, TRest>>
  : TReduce;

export function group<TObject, TKeys extends unknown[]>(
  iterable: Iterable<TObject>,
  ...keys: {
    [Index in keyof TKeys]: (
      value: TObject,
      index: number,
      values: TObject[],
    ) => TKeys[Index];
  }
): NestedInternMap<TObject, TObject[], TKeys>;

```

Reply:
@Methuselah96 I have committed new changes to this PR, taking your ideas as a basis for the refactoring. The way you expressed the types is way more efficient than my previous implementation and more readable as well 👍 I have also added new test cases for explicitly providing the type parameters to each function and tested those via the `// $ExpectType ...` syntax. Could you have a look at the changes again?

Thanks for your support on this!